design/baremetal-operator/bmo-part-of-capm3.md

```diff
@@ -0,0 +1,283 @@
+
+## Proposed structure of baremetal-operator deployment folder
+
+```diff
+tree deploy/
+
+deploy/
+├── crds
+│   ├── kustomization.yaml
+│   └── metal3.io_baremetalhosts_crd.yaml
+├── default
+│   ├── ironic_bmo_configmap.env
+│   ├── kustomization.yaml
+│   └── kustomizeconfig.yaml
+├── ironic_ci.env
+├── namespace
+│   ├── kustomization.yaml
+│   └── namespace.yaml
+├── operator
+│   ├── bmo.yaml
+│   └── kustomization.yaml
+├── rbac
+│   ├── kustomization.yaml
+│   ├── role_binding.yaml
+│   ├── role.yaml
+│   └── service_account.yaml
+└── role.yaml -> rbac/role.yaml
+
+```
+
+As we can see, the `deploy` directory has one top level folder for deployment,
+namely `default` and it deploys only baremetal-operator through kustomization
+file and uses kustomization config file for teaching kustomize where to look at
+when substituting variables. In addition, `crds`, `namespace` and `rbac` folders
+have their own kustomization and yaml files. The following directory tree
+visualizes a new structure of ironic-deployment folder in more detail:
+
+## Proposed structure of ironic-deployment folder
+
+```diff
+tree ironic-deployment/
+
+ironic-deployment/
+├── default
+│   ├── ironic_bmo_configmap.env
+│   └── kustomization.yaml
+├── ironic
+│   ├── ironic.yaml
+│   └── kustomization.yaml
+└── keepalived
+    ├── ironic_bmo_configmap.env
+    ├── keepalived_patch.yaml
+    └── kustomization.yaml
+
+```
+
+In the above provided tree, ironic-deployment folder has three top level folders
+for deployments, namely `default`, `ironic` and `keepalived`. `default` and
+`ironic` folders will deploy only ironic, whereas, `keepalived` folder deploys
+the ironic with keepalived. As the name implies, `keepalived/keepalived_patch.yaml`
+patches the default image URL through kustomization. The user should run the
+following commands to be able to meet requirements of each use case as provided
+below:
+
+### Commands to deploy "production" use case after structural changes
+
+```diff
+kustomize build $BMOPATH/deploy/default | kubectl apply -f-
+kustomize build $BMOPATH/ironic-deployment/default | kubectl apply -f-
+
+```
+
+### Command to deploy "local developer" use case after structural changes
+
+```diff
+kustomize build $BMOPATH/ironic-deployment/default | kubectl apply -f-
+
+```
+
+### Command to deploy "ironic outside of the cluster" use case after structural changes
+
+```diff
+kustomize build $BMOPATH/deploy/default | kubectl apply -f-
+
+```
+
+where $BMOPATH points to the baremetal-operator path.
+
+To this end, "production", "local developer" and "ironic outside of the cluster"
+use cases will be met, as new structure will fulfill the needs of those cases
+with cleaner and leaner organization of deployments in place.
+
+## Pivoting use case
+
+Once above outlined folder structural changes are made, we can deploy BMO on its
+own in CAPM3 and Ironic independently to meet the use case of pivoting, where
```

Comment:
Saying "deploy BMO on its own in CAPM3" is a little vague. I don't think you mean literally that the BMO code will run in the same process as CAPM3. I think you mean that there would be 1 Deployment that describes a Pod containing both BMO and CAPM3. Is that right?

Reply:
It is a kustomization file that is residing in CAPM3 referencing to bmo deployment in BMO, please check this [link](https://github.com/metal3-io/cluster-api-provider-metal3/blob/f43c80935472ae7f3550ab62f9aedcee0cecb8c9/config/bmo/kustomization.yaml#L5).

Reply:
OK. Could you please clarify the text with those details?

Reply:
Sure, I will update it accordingly as suggested, thanks.

Reply:
Done.